### ISSUE TITLE
[feature request] iOS support

### ISSUE THREAD

**Issue by arooni** [koreader/koreader#5886]
* KOReader version: whichever
* Device: iphone x

i picked up an iphone recently just to see what the hype was about.   i think i still prefer my pixel but it'd be super nice to have reading positions transfer from koreader on my paperwhite/android phone/android tablet to my iphone.  seems like this [has been brought up before on mobileread](https://www.mobileread.com/forums/showthread.php?t=275350) but the answer was support anytime soon wasn't likely.

i for one would love to see a built app in the ios app store (i'm not sure how else to install apps unless i jailbreak this phone), and would pay a reasonable price to not have to build with xcode or deal with dependencies.   any one else have this interest?  

many thanks!

**Comment by NiLuJe** [comment]
Magic 8 Ball says *extremely unlikely*.

(Can't execute dynamically generated code on iOS (that means no JIT *at all*), cross-compiling from another host than Darwin is extremely finicky, frontend would probably need to be in ObjC/Swift, .... That's a lot.)

**Comment by NiLuJe** [comment]
TL;DR: Apple can suck it.

**Comment by pazos** [comment]
@NiLuJe: you can sideload a couple (up to five I think) apps on your own device. Apps will work for 7 days on non-paid accounts and a whole year on paid dev accounts.

Porting KOReader would be possible for jailbroken devices only, where we can keep the scheme of running luajit and load (and interact) with shared libraries on demand.

**Comment by NiLuJe** [comment]
@pazos: Huh. I was just thinking/hoping that they might be doing some account shenanigans if the Apple ID was a dev one, thanks for the clarification ;).

Sidebar: is iOS jailbreaking still a thing?

**Comment by Frenzie** [comment]
> (Can't execute dynamically generated code on iOS (that means no JIT at all)

Are you saying that https://love2d.org/ and Electron apps like Skype run without JIT on iOS?

**Comment by NiLuJe** [comment]
I actually have no idea how Electron apps would even work, since even Chrome/Firefox is just a pretty wrapper around Apple's WebKit on iOS, not actually Blink/Gecko ;).

**Comment by Frenzie** [comment]
Many of those Electron apps quite literally work in the browser. Skype & Discord do, although Skype (web.skype.com) will refuse in Firefox. I seem to have made an incorrect assumption of some sort in any case, because Electron says the following:

> Compatible with Mac, Windows, and Linux, Electron apps build and run on three platforms.

So logically they use something very much like Electron, but not actually Electron. I think Discord might be using React Native, in which case Electron is merely React Native's way of doing desktop? Maybe Skype does the same. But if it has to depend on system WebKit on iOS I guess life is more annoying for those devs in all kinds of ways, but ultimately quite possibly better for us users. (Because of a certain degree of cross-browser compatibility; Blink and WebKit have diverged quite a bit.)

**Comment by arooni** [comment]
I would totally jailbreak my iphone for the ability to run koreader.  No brainer.  Sounds like the amount of dev work to get it working would be intense tho :\

**Comment by pazos** [comment]
> Sidebar: is iOS jailbreaking still a thing?

Yup, IMO is the single thing between an iPad and a toyPad. The hability to have a proper file manager on device and mount arbitrary dirs on my linux machine when the device is unlocked is a must for my usage.

Also, you can fakesign your own apps and avoid the 7days restrictions on free accounts. Some apps are only available for jailbroken devices: mostly emulators and kodi.

**Comment by pazos** [comment]
Closing this one. Pull Requests are welcome!

**Comment by gingerbeardman** [comment]
I'm not requesting anything but thought it would be good to update this issue with more recent info:

You can run Lua in an app, and you can run code that didn't exist before:
- https://apps.apple.com/gb/app/touch-lua/id525273327
- https://apps.apple.com/gb/app/lua-ide/id1549382090

In fact, Apple run Lua code in places as part of iOS, since iOS 12 in 2018:
- https://twitter.com/_inside/status/1026173832527265792

Files (books, plugins,  can be added through shared folder (using Finder's device management window) or the app can provide an iCloud folder. 

Outstanding questions:
- annual developer fee (group fund?)
- sideloading (forthcoming EU law?)
- competing with built-in Books app?

**Comment by pazos** [comment]
Hi, no iOS will never happen. ffi and jit are not available. 
Lets move on :)

**Comment by Frenzie** [comment]
JIT isn't that big a problem (on a relatively fast device like an iPhone), but no FFI would obviously be impossible.

**Comment by pazos** [comment]
> JIT isn't that big a problem (on a relatively fast device like an iPhone), but no FFI would obviously be impossible.

the lack of JIT is what makes FFI unusable:  https://twitter.com/LuaJIT/status/208585095848796160
https://www.freelists.org/post/luajit/LuaJit-FFI-with-iOS,1

**Comment by Frenzie** [comment]
Given those links I think it'd be sufficiently usable. You can see for yourself with `luajit -j off reader.lua` on Kobo or something for a worst case scenario, and an iPhone will be much faster. But crucially, I'm not interested in doing an iOS port. ;-)

**Comment by gingerbeardman** [comment]
That is pretty crucial :)

Anyway, good to see some modern chat in this thread. I'll remain subscribed to it (though I use an Android and old Sony Reader PRS-650 for my reading)

**Comment by NiLuJe** [comment]
Given how much of a PITA even barely maintaining the macOS port is, I'd go so far as even *veto* anyone actively working on this, because it would inevitably lead to insane bitrot.

**Comment by gingerbeardman** [comment]
I would surely get a lot of users.

**Comment by NiLuJe** [comment]
Users are not maintainers ;).

**Comment by gingerbeardman** [comment]
So true.

If it ever happens, big if, I'd offer to help maintain it.

**Comment by thatsthat** [comment]
Does the new EU law for iOS have the potential to make this any easier?

https://support.apple.com/en-gb/118110

**Comment by Frenzie** [comment]
That only affects the "putting it on the App Store" part, which I doubt was too much of an obstacle.

**Comment by thatsthat** [comment]
One more good reason for not buying an iPhone then 😃

**Comment by NebulaBC** [comment]
JIT is very possible on iOS and has been. You can use jitstreamer or sidestore jit, or altstore jit though that’s local network only. It shouldn’t be a show stopper, though I know that iOS apps are borderline impossible to maintain, especially if they’re not your main target. Like a lot of people, it sure would be neat to read on my iPad though.

**Comment by pazos** [comment]
> JIT is very possible on iOS and has been. You can use jitstreamer or sidestore jit, or altstore jit though that’s local network only

Traditionaly on jailbroken devices, good to know it is possible sideloading. Anyhow, we were talking about https://github.com/LuaJIT/LuaJIT/issues/1072#issuecomment-1706201991

> especially if they’re not your main target

That's the key factor. KOReader is easily portable across linux embedded devices. Porting to other platforms means writting a bunch of glue code to adapt the platform to KOReader expectations.

That was done for android and for the desktop/sdl, which are "open" platforms where anybody with a computer can develop.

The fact that we have a macos target means sdl + handle a few quirks. A new iOS target would mean something like https://github.com/koreader/android-luajit-launcher but just for iOS. Nobody here wants to hear about that :D. 

> Like a lot of people, it sure would be neat to read on my iPad though.

The future solution is https://github.com/readest/readest, built by one of the guys that started KOReader, @chrox. It plans to support KOReader syncing and uses a stack based on a browser engine that makes it easily portable across regular OSes.

In the mean time if somebody wants to attempt to port KOReader to iOS in their free time nobody here will complain as long as it is a standalone project that we're not forced to maintain.

**Comment by scj643** [comment]
> The future solution is https://github.com/readest/readest

The future is now

**Comment by hezi** [comment]
Don’t know if anyone still cares about this, but I really just wanted koreader on my iPhone and iPad, so here it is:
https://github.com/hezi/koreader-ios

Didn’t test everything, so might still be rough around the edges, but the main features work. Let me know if you find it useful, I’ve been using it a lot to sync my reading between my phone/tablet and eink devices with no issues.

**Comment by gingerbeardman** [comment]
Thanks @hezi Jorge!
